gmail/gmail-connector/gmail-connector-2.0.0/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/GmailIMAPClientLoader.java

```diff
@@ -0,0 +1,131 @@
+                        .getProperty(GmailConstants.GMAIL_IMAP_STORE_INSTANCE);
+
+        // Use if there exists an already stored IMAPStore instance.
+        if (prestoredInstance != null) {
+            log.info("Retriving the prestored IMAPstore instance");
+            return (IMAPStore) prestoredInstance;
+        }
+
+        // Login mode should have been defined during either "init" or
+        // "passwordLogin" operations.
+        Object loginMode = axis2MsgCtx.getProperty(GmailConstants.GMAIL_LOGIN_MODE);
+        if (loginMode == null) {
+            String errorLog = "Gmail configuration details were not initialized";
+            log.error(errorLog);
+            ConnectException connectException = new ConnectException(errorLog);
+            throw (connectException);
+        }
+
+        IMAPStore store = null;
+
+        // Perform SASL authentication if configured using the "Password Login"
+        // operation.
+        if (loginMode.toString().equals(GmailConstants.GMAIL_SASL_LOGIN_MODE)) {
+            log.info("SASL authentication starts");
+            try {
+                store =
+                        GmailSASLAuthenticator.connectToIMAP(messageContext.getProperty(GmailConstants.GMAIL_USER_USERNAME)
+                                        .toString(),
+                                messageContext.getProperty(GmailConstants.GMAIL_USER_PASSWORD)
+                                        .toString());
+            } catch (MessagingException e) {
+                log.error("Failure in SASL authentication");
+                throw (e);
+            }
+        }
+
+        // Perform OAuth authentication if configured using the "init"
+        // operation.
+        else if (loginMode.toString().equals(GmailConstants.GMAIL_OAUTH_LOGIN_MODE)) {
+            if (axis2MsgCtx.getProperty(GmailConstants.GMAIL_OAUTH2_PROVIDER) == null) {
+                log.info("Initializing OAuth2 provider");
+                GmailOAuth2SASLAuthenticator.initializeOAuth2Provider();
+                axis2MsgCtx.getOperationContext().setProperty(GmailConstants.GMAIL_OAUTH2_PROVIDER,
+                        "Initialized");
+            }
+
+            log.info("OAuth2 authentication starts");
+            try {
+                store =
+                        GmailOAuth2SASLAuthenticator.connectToIMAP(messageContext.getProperty(GmailConstants.GMAIL_OAUTH_USERNAME)
+                                        .toString(),
+                                messageContext.getProperty(GmailConstants.GMAIL_OAUTH_ACCESS_TOKEN)
+                                        .toString());
+            } catch (MessagingException e) {
+                log.error("Failure in OAuth2 authentication.");
+                throw (e);
+            }
+
+        } else {
+            String errorLog = "Gmail configuration details were not initialized";
+            log.error(errorLog);
+            ConnectException connectException = new ConnectException(errorLog);
```

Comment:
This logic could have been made much simpler I've already suggested in one of my previous comments.